### DISCUSSION TITLE
Is there a way to OneHotEncode an entire array?

### DISCUSSION THREAD

**Discussion by moisestohias** [mrdbourke/pytorch-deep-learning#149] [Q&A] [answered]
Hi,
I wonder, Is there a way to OneHotEncode an entire array? 
```python
def OneHot(y, NumOfClasses):
    e = np.zeros((NumOfClasses, 1))
    e[y] = 1.0
    return e
```
Basically obviating doing this `np.array([OneHot(i, 10) for i in y]),`
I tried using `np.vectorize` function, but that didn't work because of indexing issues I guess, I am not sure. 
Anyway I would appreciate you suggests.

**Comment by mrdbourke** [comment, marked as answer]
Hi @moisestohias,

You may want to try Scikit-Learn's OneHotEncoder - http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html

**Comment by moisestohias** [comment]
I have found many solutions on the Internet 
Different way to one hot
#### Resources:
* [stackoverflow:1](https://stackoverflow.com/questions/29831489/convert-array-of-indices-to-one-hot-encoded-array-in-numpy)
* [stackoverflow:2](https://stackoverflow.com/questions/36960320/convert-a-2d-matrix-to-a-3d-one-hot-matrix-numpy)

__!Note: for one hot encoding you can use scipy.sparse as well.__
__!TIP: don't rely on the a.max()+1 to get the number of classes.__

__Note: using eye is the only useful solution for an input N-D matrix to one-hot N+1D matrix. __
Example: 
```python
input_matrix=np.asarray([[0,1,1] , [1,1,2]]) ; np.eye(3)[input_matrix] # output 3D tensor
def one_hot(a, classes=10): return np.eye(classes)[a] # .T Transpose if you don't want torch style

def one_hot(a: np.ndarray):
  # a.max()+1 automatically determin the number of classes based on the highest value.
  hot = np.zeros((a.size, a.max()+1), dtype=np.float32)
  hot[np.arange(a.size),a] = 1.0
  return hot

def one_hot(a, classes=10 ):
  hot = np.zeros((a.size, classes), dtype=np.float32)
  hot[np.arange(a.size),a] = 1.0
  return hot
```
